### PR TITLE
move `torch.cuda.set_device()` to enable collective calls earlier in setup

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Unblocks #8017 who is calling `trainer.log_dir` early in `Callback.setup()`, which internally leads to a broacast, but hangs because device is not yet set!

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/gpu.py [hunk 1]
```diff
@@ -26,16 +26,19 @@
 class GPUAccelerator(Accelerator):
     """ Accelerator for GPU devices. """
 
+    def setup_environment(self) -> None:
+        super().setup_environment()
+        if "cuda" not in str(self.root_device):
+            raise MisconfigurationException(f"Device should be GPU, got {self.root_device} instead")
+        torch.cuda.set_device(self.root_device)
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/gpu.py:33]: reviewer: moving the call from `Accelerator.setup()` to here. Why? 

`setup_environment()` runs earlier in the `trainer._run()` and enables us to call broadcast etc. in a hook between, for example `Callback.setup()`

setting the devices is required for collective calls to work properly.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by SeanNaren — APPROVED** [review]

**Comment by carmocca** [comment]
Do we also want to move?:

```python
pytorch_lightning//plugins/training_type/horovod.py:136:            torch.cuda.set_device(self.root_device)
pytorch_lightning//plugins/training_type/ddp_spawn.py:333:            torch.cuda.set_device(self.root_device)
```

**Comment by awaelchli** [comment]
@carmocca the one in ddp_spawn cannot be moved, it needs to set the device in the spawned subprocesses too.
for horovod I'm not confident enough to change it right now, as we have very little testing around that plugin.

**Pull request merged**